### DISCUSSION TITLE
Hierarchical Areas, Project, Sections and Tasks

### DISCUSSION THREAD

**Discussion by BeatLink** [alainm23/planify#708] [Ideas]
This issue serves to consolidate all the existing feature requests for hierarchical areas, projects, sections and/or tasks, as well as to open design discussions on which of the four to nest.

Some existing issues that request this include: 
* https://github.com/alainm23/planner/issues/416
* https://github.com/alainm23/planner/issues/169
* https://github.com/alainm23/planner/issues/255

**Comment by BeatLink** [comment]
I believe the best way to solve the issue of deciding which to nest is to follow todoist's design structure, since planner is inspired by and designed to be compatible with it. Following Todoist's design structure would also eliminate the need to reinvent the wheel
So that would mean:
* Nesting projects to 3 indentation levels,
* Nesting tasks to 4 indentation levels, 
* Keeping sections as is
* Removing areas and replacing them with the nested projects. 

It seems that this structure has also been requested by others: https://github.com/alainm23/planner/issues/371

It also appears that todoist treats subtasks as full featured tasks (including adding labels, priorities, due dates and reminders) so we could do the same. In the UI design, we can remove the subtask remove button in favor of a menu button that presents these options, or we can present the options side by side to the right of the subtask

Some links about todoists design structure:
* https://get.todoist.help/hc/en-us/articles/115001862065-Create-a-sub-project
* https://get.todoist.help/hc/en-us/articles/360003788739-Sections
* https://get.todoist.help/hc/en-us/articles/206432369-Sub-tasks

**Reply by beedell-roke** [reply]
> Nesting tasks to 4 indentation levels,

@BeatLink, that wouldn't be sufficient for me. My CalDAV tasks can be nested 20 deep.

**Comment by wada3n** [comment]
Actually we need this @alainm23 , especially sub-task for a sub-task.
I have a project on Todoist with this feature and when syncing via Planner, I found nothing inside it, only the task and the sub-task on it, while all the internal sub-tasks were gone.

**Comment by BeatLink** [comment]
I was thinking as an alternative way to represent subtasks, you could move the checkboxes from inside the task and instead move all subtasks below the parent task but just indented, similar to gnome todo. So instead of this:

![image](https://user-images.githubusercontent.com/47278150/98444330-09d9da00-20df-11eb-92d1-cd4f1a50ba04.png)

It would look more like this:

![image](https://user-images.githubusercontent.com/47278150/98444287-d008d380-20de-11eb-9c52-6ae0c5de2705.png)

So subtasks would be outside the parent task box and indented below it instead.

This would also be much simpler to implement as you could use the existing task UI for every task, parent or subtask, and all you would need is to edit the indentation level and group the tasks so the subtasks are always moved with the parent. Nested tasks would be a breeze since all you would need to do UI wise is to increase the indentation

**Comment by alainm23** [comment]
@BeatLink It's a good solution, I'll take some time to research how to do it, thank you for your time.

**Comment by BeatLink** [comment]
No problem

**Comment by rzelnik** [comment]
Once I had an idea to create a task planner that would have recursive tasks. The point behind it is that on size doesn't fit for all. Some projects are too complex for any fixed number of sublevels, while most projects don't need any sublevels. Sometimes you need just one more level of nesting and it just isn't available. Recursive nesting would be easier to implement than the whole hierarchy of projects > tasks > subtasks > second level subtasks etc., and the user could use the number of sublevels which is appropriate for a specific project.

**Comment by BeatLink** [comment]
Hmm, fair enough. I do agree that arbitrary limits on nesting are a problem. But i believe the current design is to make it compatible with todoist. How would nested tasks beyond todoist's limit be handled by the app? Would the data be available on planner only?

**Comment by BeatLink** [comment]
Alternatively, we could focus more on planner's upcoming CalDAV support and have todoist have a smaller role in planner's design

**Comment by ttsesm** [comment]
I also find it useful to add more than one sub-task layers into tasks including all the relevant features that someone finds available at the moment only in the root task objects i.e. notes, labels, priorities, progress bar etc.

**Comment by BeatLink** [comment]
Hmm. After further thought, I think following the Todoist methodology would be best. However, we need not limit ourselves in the number of nested tasks or projects. We could simply check if that specific project or task is synced with a todoist account and if so, limit the number of nesting to the todoist specifications, such as by disabling the ability to add more subtasks or projects beyond the limit. 

For local or caldav based tasks we wouldnt have that limit. I think that gives the best of both worlds.

**Comment by koudi** [comment]
> Alternatively, we could focus more on planner's upcoming CalDAV support and have todoist have a smaller role in planner's design

I, for one, don't need sub projects (I'm fine with sections) or subtasks (checklists are ok for me), but I really miss areas lacking in todoist (and thus planner) design. 

Since you're discussing possibility of different backends, is it necessary to constraint yourself to limits of todoist? How important feature is it anyway? I'm asking honestly, because I don't know and I think planner looks like an awesome software that could stand on its own.

**Comment by BeatLink** [comment]
Well Subprojects serve the same functions as areas but better. You can nest them infinitely, add tasks to them and set due dates for them. In addition you can give them descriptions and specific colours. Its a win win.

**Comment by koudi** [comment]
Yes, they're technically similar, but conceptually they are different. Areas have no due dates, they are never really 'done' and they have no meaning of progress (as in x/y tasks done)

**Comment by BeatLink** [comment]
Hmm. Fair enough. But couldnt the same be achieved with a user created project that has no due date?

Nested subprojects would help make planner more user customizable by allowing them the freedom to create their own workflow. Rather than constraining them to areas, we could simply give them nested subprojects and allow the user to determine what those projects mean to them. In your case (as mine) the top level projects could be simple categorizations without an expected due date

I make that distinction through project names. The top level ones i save as nouns that describe the category (eg Home) and under them i create projects with more active language in the title (eg Renovate kitchen)

**Comment by koudi** [comment]
Yes, pretty much the same thing could be achieved with subprojects. In the end this this more of a UX thing rather than features (at least for me).

Just an idea: what if we had a project flag "this is area", which would give it slightly different treatment - hiding progress, due dates and possibly removing first level of nesting for it's sub-projects (for example like in [Things](https://culturedcode.com/things/whats-new/))? This way syncing to todoist would still work as normal project, but with these cosmetic enhancements on the planner side.

**Comment by BeatLink** [comment]
I like it. It would be up to @alainm23 though.

**Comment by alainm23** [comment]
I'm currently working on subtasks, in the same way as Todoist or Gnome Todo works.
@koudi It's an interesting idea, I would like to talk more about it and the advantage over a project itself.

**Comment by koudi** [comment]
Honestly I don't like subtasks the todoist way, I think that the simple checkbox list is great and simple.

As for areas...go ahead, ask :). There aren't any actual advantages over projects (as it would still be project under the hood), I just think it's nicer way to organise stuff as I described above with slightly different semantics.

**Comment by BeatLink** [comment]
I think this issue would be best served as a discussion instead

**Comment by BeatLink** [comment]
It would be really cool if tags could be hierarchical as well. It would allow greater flexibility of categorization and tagging. Getting Things GNOME has this feature already if you want to take a look at it in practice

**Reply by koudi** [reply]
I think this would be unnecessary over-complication. Getting things gnome does indeed have hierarchical tags, but it doesn't have (sub)projects and organizing tasks is done purely via these tags.

**Reply by BeatLink** [reply]
Yeah, that makes sense
